### PR TITLE
[SPARK-31428][SQL][DOCS] Document Common Table Expression in SQL Reference

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Document Common Table Expression in SQL Reference

### Why are the changes needed?
Make SQL Reference complete

### Does this PR introduce any user-facing change?
Yes
Screen Shot 2020-04-13 at 12 06 35 AM

Screen Shot 2020-04-13 at 12 07 09 AM

Screen Shot 2020-04-13 at 12 07 35 AM

Also link to Select page

Screen Shot 2020-04-12 at 4 14 30 PM

### How was this patch tested?
Manually build and check

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @maropu

File: docs/sql-ref-syntax-qry-select-cte.md [hunk 1]
```diff
@@ -19,4 +19,120 @@ license: |
...
+SELECT * FROM t2;
+  +---+
+  |  2|
+  +---+
+  |  2|
+  +---+
+
+-- SET spark.sql.legacy.ctePrecedencePolicy = LEGACY (the behavior in version 2.4 and earlier)
+-- outer CTE definitions take precedence over inner definitions.
+SET spark.sql.legacy.ctePrecedencePolicy = CORRECTED
```
Comment (viirya) [inline docs/sql-ref-syntax-qry-select-cte.md:117]: `SET spark.sql.legacy.ctePrecedencePolicy = LEGACY`

Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-cte.md]: Fixed. Thanks!

File: docs/sql-ref-syntax-qry-select-cte.md [hunk 2]
```diff
@@ -19,4 +19,120 @@ license: |
   limitations under the License.
 ---
 
-**This page is under construction**
+### Description
+
+A common table expression (CTE) defines a temporary result set that a user can reference within the duration of a SQL statement. A CTE can be used wherever a SELECT statement is used; for example, it can be used in a SELECT, INSERT, UPDATE, DELETE, MERGE or CREATE VIEW statement. A CTE has a name with an optional column names list, followed by a query expression. When a CTE references itself, it is called a recursive CTE.
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-cte.md:24]: nit: How about this?
`A common table expression (CTE) defines a temporary result set that a user can reference within the duration of a SQL statement. `
=>
`A common table expression (CTE) defines a temporary result set that a user can reference possibly multiple times within the scope of a SQL statement. `

Comment (maropu) [reply docs/sql-ref-syntax-qry-select-cte.md:24]: Is this statement necessary? `A CTE can be used wherever a SELECT statement is used; for example, `? How about just saying `within the duration of a SQL statement. A CTE can be used in SELECT, INSERT, UPDATE, DELETE, MERGE, or CREATE VIEW statements`?

Comment (maropu) [reply docs/sql-ref-syntax-qry-select-cte.md:24]: UPDATE and MERGE are supported? https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala#L765-L768

File: docs/sql-ref-syntax-qry-select-cte.md [hunk 3]
```diff
@@ -19,4 +19,120 @@ license: |
...
+
+### Syntax
+
+{% highlight sql %}
+WITH common_table_expression [ , ... ]
+{% endhighlight %}
+
+While ```common_table_expression``` is defined as
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-cte.md:32]: a single backquote? https://github.com/apache/spark/blame/master/docs/sql-ref-syntax-qry-select.md#L38

File: docs/sql-ref-syntax-qry-select-cte.md [hunk 4]
```diff
@@ -19,4 +19,120 @@ license: |
...
+        SELECT * FROM t
+    )
+SELECT * FROM t2;
+  +---+
+  |  2|
+  +---+
+  |  2|
+  +---+
+
+-- SET spark.sql.legacy.ctePrecedencePolicy = LEGACY (the behavior in version 2.4 and earlier)
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-cte.md:115]: The 3.0 document needs to describe this legacy behaivour? I personally think we don't need it. cc: @gatorsmile

File: docs/sql-ref-syntax-qry-select-cte.md [hunk 5]
```diff
@@ -19,4 +19,120 @@ license: |
...
+    )
+SELECT * FROM t2;
+  +---+
+  |  1|
+  +---+
+  |  1|
+  +---+
+
+-- CTE should not fall into infinite loop by referencing self
+WITH s AS (SELECT 1 FROM s) SELECT * FROM s;
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-cte.md:132]: > When a CTE references itself, it is called a recursive CTE.

Why do you think we need to describe this unsupported behaivour?

File: docs/sql-ref-syntax-qry-select-cte.md [hunk 6]
```diff
@@ -19,4 +19,120 @@ license: |
...
+  Error in query: Table or view not found s; line 1 pos 25;
+{% endhighlight %}
+
+### Related Statement
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-cte.md:136]: nit: currently, we have the single entry in this section though, I think `### Related Statement"s"` is okay for consistency.

File: docs/sql-ref-syntax-qry-select-cte.md (hunk 2 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-cte.md:24]: I guess maybe only mention the SELECT?
```A common table expression (CTE) defines a temporary result set that a user can reference possibly multiple times within the scope of a SQL statement. A CTE is used mainly in a SELECT statement. It is defined with a name including an optional column names list, followed by a query expression. When a CTE references itself, it is called a recursive CTE.```
Does this look OK to you?

File: docs/sql-ref-syntax-qry-select-cte.md (hunk 5 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-cte.md:132]: I will remove

File: docs/sql-ref-syntax-qry-select-cte.md (hunk 2 above)
Comment (maropu) [reply docs/sql-ref-syntax-qry-select-cte.md:24]: We need to describe `It is defined with a name including an optional column names list, followed by a query expression.` in the head section? [The syntax section](https://github.com/apache/spark/pull/28196/files#diff-771ee2387e08ce855bea13e43117c735R26-R35) looks enough to me.

File: docs/sql-ref-syntax-qry-select-cte.md [hunk 7]
```diff
@@ -19,4 +19,120 @@ license: |
...
+  |  1|
+  +---+
+  |  1|
+  +---+
+
+-- CTE should not fall into infinite loop by referencing self
+WITH s AS (SELECT 1 FROM s) SELECT * FROM s;
+  Error in query: Table or view not found s; line 1 pos 25;
+{% endhighlight %}
+
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-cte.md:135]: How about adding one more CTE example in `CREATE VIEW`?

File: docs/sql-ref-syntax-qry-select-cte.md (hunk 2 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-cte.md:24]: will remove.

File: docs/sql-ref-syntax-qry-select-cte.md (hunk 7 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-cte.md:135]: Will do

File: docs/sql-ref-syntax-qry-select-cte.md (hunk 4 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-cte.md]: removed legacy behavior

File: docs/sql-ref-syntax-qry-select-cte.md [hunk 8]
```diff
@@ -19,4 +19,111 @@ license: |
...
+    WITH t(a, b, c, d) AS (SELECT 1, 2, 3, 4)
+    SELECT * FROM t;
+SELECT * FROM v;
+  +---+---+---+---+
+  |  a|  b|  c|  d|
+  +---+---+---+---+
+  |  1|  2|  3|  4|
+  +---+---+---+---+
+
+-- Name conflict in nested CTE. Spark throws an AnalysisException by default.
```
Comment (viirya) [inline docs/sql-ref-syntax-qry-select-cte.md:109]: -- While name conflict is detected in nested CTE, AnalysisException is thrown by default.

**Comment by HyukjinKwon** [comment]
FYI @peter-toth

File: docs/sql-ref-syntax-qry-select-cte.md (hunk 8 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-cte.md]: Fixed. Thanks!

File: docs/sql-ref-syntax-qry-select-cte.md [hunk 9]
```diff
@@ -19,4 +19,111 @@ license: |
...
+    WITH t(a, b, c, d) AS (SELECT 1, 2, 3, 4)
+    SELECT * FROM t;
+SELECT * FROM v;
+  +---+---+---+---+
+  |  a|  b|  c|  d|
+  +---+---+---+---+
+  |  1|  2|  3|  4|
+  +---+---+---+---+
+
+-- If name conflict is detected in nested CTE, then AnalysisException is thrown by default.
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-cte.md:109]: We still need to describe this? I personally think users can see the migration guide for these kinds of legacy behaviours.

**Review by maropu — APPROVED** [review]
LGTM except for one comment.

File: docs/sql-ref-syntax-qry-select-cte.md (hunk 9 above)
Comment (huaxingao) [reply docs/sql-ref-syntax-qry-select-cte.md:109]: Thanks for your comment.
I prefer to document the default behavior. I actually thought the default behavior is "inner CTE definitions take precedence over outer definitions". When I tried the example, I was surprised to see the Exception, then I looked the code and found out I need to set spark.sql.legacy.ctePrecedencePolicy to make it work. So I guess it is worth mentioning the default here.

**Comment by huaxingao** [comment]
cc @srowen

**Review by srowen — APPROVED** [review]
I'm OK with a little over documentation here

**Comment by maropu** [comment]
Thanks! Merged to master/3.0.

**Comment by huaxingao** [comment]
Thanks, all!

**Pull request closed without merging**